core/mvc/models/behaviors/shortCodeBehavior.php

```diff
@@ -0,0 +1,152 @@
+     * This function is called before a model is created.
+     *
+     * @param  ModelInterface $model
+     * @return void
+     */
+    protected function beforeValidationOnCreate(&$model)
+    {
+        $model->shortCode = $this->generateShortCode($model->name);
+    }
+
+    /**
+     * This function generates a short code based on the project name.
+     *
+     * @param  string $name
+     * @return string
+     */
+    public function generateShortCode($name)
+    {
+        if (strpos($name, ' ') !== false) {
+            return $this->generateShortCodeFromMultipleWords($name);
+        } else {
+            return $this->generateShortCodeFromSingleWord($name);
+        }
+    }
+
+    /**
+     * Generate a short code from multiple words.
+     *
+     * @param  string $name
+     * @return string
+     */
+    protected function generateShortCodeFromMultipleWords($name)
+    {
+        $words = explode(' ', $name);
+        $shortCode = '';
+
+        foreach ($words as $word) {
+            $shortCode .= substr($word, 0, 1);
+        }
+
+        $shortCode = $this->updateShortCodeIfExistsInDatabase($shortCode);
+
+        return strtoupper($shortCode);
+    }
+
+    /**
+     * Generate a short code from a single word.
+     *
+     * @param  string $name
+     * @return string
+     */
+    protected function generateShortCodeFromSingleWord($name)
+    {
+        $shortCode = strtoupper(substr($name, 0, 5));
+
+        // Check if the short code exists in the database
+        $shortCode = $this->updateShortCodeIfExistsInDatabase($shortCode);
+
+        return $shortCode;
+    }
+
+    /**
+     * Check if the short code exists in the database. If exists then append a unique number at the
+     * last of the shortcode.
+     *
+     * @param  string $shortCode
+     * @return string
+     */
+    protected function updateShortCodeIfExistsInDatabase($shortCode)
+    {
+        $searchQuery = "%$shortCode%";
+        $shortCodes = [];
+        $number = '';
+
+        $projects = $this->getProjectsByShortCode($searchQuery);
+
+        foreach ($projects as $project) {
+            $shortCodes[] = $project->shortCode;
+        }
+
+        $numbers = [];
+
+        foreach ($shortCodes as $item) {
+            if (preg_match('/\d+/', $item, $matches)) {
+                $numbers[] = (int)$matches[0];
+            } else {
+                $numbers[] = 0;
+            }
+        }
+
+        if (!empty($numbers)) {
+            $number = max($numbers);
+            $number++;
+        }
+
+        $updatedShortCode = "$shortCode$number";
+        return $updatedShortCode;
+    }
+
+    /**
+     * Retrieve projects by short code from database.
+     *
+     * @param  string $shortCode
+     * @return \Phalcon\Mvc\Model\ResultsetInterface
+     */
+    public function getProjectsByShortCode($shortCode)
+    {
+        return \Gaia\MVC\Models\Project::find(
+            [
+            "conditions" => "shortCode LIKE :shortCode:",
```

Comment:
Shouldn't this be shortCode: and not :shortCode:

Reply:
No, **:shortCode:** is the bind parameter for the condition.

Reference:
https://docs.phalcon.io/3.4/db-models/#binding-parameters